src/main/groovy/com/lesfurets/jenkins/unit/PipelineTestHelper.groovy

```diff
@@ -23,28 +23,100 @@ class PipelineTestHelper {
     protected static Method SCRIPT_SET_BINDING = Script.getMethod('setBinding', Binding.class)
 
     /**
-     * Simple container for holding mock script output.
+     * Simple container for handling mocked scripts (`bat`, `sh`, etc).
      */
-    class MockScriptOutput {
+    class MockScriptHandler {
+        // optional filter used to determine whether a script matches this mock
+        String filter = null
+
+        // mocked script results, specified as either value or callback
         String stdout = null
         int exitValue = -1
         Closure callback = null
 
-        MockScriptOutput(String stdout, int exitValue) {
+        MockScriptHandler(String filter, String stdout, int exitValue) {
+            this.filter = filter
             this.stdout = stdout
             this.exitValue = exitValue
         }
 
-        MockScriptOutput(Closure callback) {
+        MockScriptHandler(String filter, Closure callback) {
+            this.filter = filter
             this.callback = callback
         }
+
+        /**
+         * match a script invocation against our filter
+         */
+        private boolean match(String script) {
+            // if no filter is set, this matches everything
+            if (filter == null) {
+                return true
+            }
```

Comment:
Check out the three `runShWithoutMockOutput*()` tests, they all assert the behaviour when no mocks are configured. Those three tests pass without modification.

What I changed was that you now have the possibilty to adjust this behaviour by using a `null` filter, if that makes sense for your test. Also, it's documented in the last two sentences:

>     Extend nested helper class.
>     
>     Instead of just holding some data, encapsulate both matching a call and
>     executing a mock operation. Further, instead of a map, use a stack of
>     handlers.
>     
>     This mostly reorders some code, but also prepares for patterned call
>     matching. As a first addition, it allows you to configure a fallback
>     handler (`pattern = null`) that matches everything. This is useful if
>     you want to detect unexpected calls.

Reply:
Actually, I'm mistaken about the previous behavior. I could have _sworn_ that we raised if no script could be found, but I think that was the super-old behavior of `addShMock` before it was merged with this repo (it used to be part of an add-on library developed by my company which we eventually merged with this one).

Anyways, as you are of course aware, until https://github.com/jenkinsci/JenkinsPipelineUnit/pull/483 the default behavior for unregistered script outputs was to return `\nbbb\nccc\n`. So I was incorrect, this isn't really a breaking API change.